[spec] Add a spec for session-scoped connections.

sfc-gh-jkinkead opened the pull request:
Rendered spec: https://issues.streamlit.app/spec_renderer?pr=13037

Add a spec for session-scoped connections.

This spec also includes session-scoped cached resources, and support for Snowflake restricted caller's rights sessions while running on Snowpark Container Services.

---

**Contribution License Agreement**

By submitting this pull request you agree that all contributions to this project are made under the Apache 2.0 license.

sfc-gh-jkinkead committed:
Add a spec for session-scoped connections.

This spec also includes session-scoped cached resources, and support for Snowflake restricted caller's rights sessions while running on Snowpark Container Services.

sfc-gh-jkinkead committed:
Fix typos.

jrieke commented:
Added a quick app to render this spec in Markdown (as well as other submitted specs): https://issues.streamlit.app/spec_renderer

jrieke commented:
This is awesome, thanks! I love adding these features, especially since they were also requested from the community a bunch of times. Few points of feedback below. 

### Parameter naming:

- **`session_scoped: bool = False` on `st.cache_resource`**: I think we should do this as `scope="global"|"session"` instead. I think that's a) a bit easier to understand, b) has precedent since we're already using `scope` in `st.rerun`, and c) if we wanted to, we could extend it in the future, e.g. if we'd want to have something like `scope="user"` (not sure if we'll ever do that, but would be good to keep it open). 
- **`eviction_hook` on `st.cache_resource`**: A few options that would be easier to understand and more in line with other parameters (especially `on_change`):
   - `on_teardown`
   - `on_cleanup`
   - `on_close`
   - `on_shutdown`
   - `on_clear`

    We could also do all of those without the `on`. Don't have a favorite yet, curious to hear what others think!
   
- Not an extremely strong opinion but I *think* on `st.connection` we should use the same parameter names as on `st.cache_resource`. I guess we will just pass the values through to `st.cache_resource` anyway, right? I.e. the `shutdown` function you propose above is not different from the `eviction_hook` function, correct? In that case, I think it creates less confusion just using the same names (and we should find a name that works both for `st.cach_resource` and `st.connection`!). 

### Behavior questions

- "This will NOT be called when an app is shut down for global resources." -> This means when I kill the Streamlit server the `eviction_hook` function will not be called? Does this create any problems, e.g. for a database connection, because something on the side of the database is kept open? Or is killing the Python process anyway equivalent to shutting down the connection?
- The proposed `snowflake-rcr` connection:
    - Is that something that's usable outside of a SiS app? 
    - Would the dev explicitly call `st.connection("snowflake-rcr")` in their SiS app, or are we calling this automatically if the app is using caller's rights?
    - I think what I'm trying to get to: is this something we want to have public and distributed as part of the Streamlit repo, or is this something that should be private and potentially not even ship with Streamlit, since it's a SiS-only feature?
   - If it should be public, we should also think about renaming it (e.g. `snowflake-callers-rights`) since we usually try to avoid acronyms in Streamlit. Or should this even be a parameter on the existing `st.connection("snowflake")`?

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
````diff
@@ -0,0 +1,170 @@
+
+```python
+@property
+def is_session_scoped(self) -> bool:
````
Do we need to expose `is_session_scoped` as a public function of what is returned in `st.connection` or is this just an implementation detail for new connections? In that case we could just make it internal (prefix with underscore).

sfc-gh-jkinkead replied:
I feel like this is useful information for consumers when they get a connection, and I wouldn't necessarily hide this from them. It seems unlikely to change.

There would be cases where knowing the scope could be useful, especially if we wanted to integrate this into caching someone (like warning people when using a session-scoped connection while inside a cache decorator which was globally-scoped).

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,170 @@
+    """
+    return False
+
+def shutdown(self) -> None:
```
Same as the question above, is this something that a user would also eventually want to call on the `st.connection` object or is this an internal detail to implement new connections?

sfc-gh-jkinkead replied:
This would be useful for connections used outside of `st.connection`. I don't know if that's a normal use-case, though.

lukasmasuch replied:
One thought on `shutdown`: Maybe its better to call this `close` since this is the common name in Python for closing any type of resource (file, connection, ...). I guess if a user calls this, its will close the underlying connection.

sfc-gh-jkinkead replied:
WFM. Renamed.

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,170 @@
+    library as well.
+* Add cache eviction hooks. If we’re building shutdown hooks for connections, we should likely just build these as
+    eviction hooks on `cache_resource`. These are useful for any resource cache with a TTL or max entry count - and
+    they’re semi-mandatory for session-scoped cached resources.
```
The feature request here seems to suggest that there might be also some value for an eviction hook on a cache function without ttl/max entries: https://github.com/streamlit/streamlit/issues/8674 However, I don't really understand when this should be called without ttl/max entries/session scoped since the cached object will never get evicted during the lifecycle of the app.

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,170 @@
+    self,
+    # Existing args omitted for clarity.
+    session_scoped: bool = False,
+    eviction_hook: EvictionHook | None = None
```
My favourites regarding API naming: `scope="global" | "session"` and `on_expire`.

sfc-gh-jkinkead replied:
I went with these names.

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
````diff
@@ -0,0 +1,170 @@
+    # Note that this default implementation is a no-op.
+```
+
+These two new methods will be passed to the `cache_resource` call in `st.connection`, using the new new parameters.
````
```suggestion
These two new methods will be passed to the `cache_resource` call in `st.connection`, using the new parameters.
```

sfc-gh-jkinkead replied:
Fixed.

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,170 @@
+
+This new connection type will be registered as `snowflake-rcr`.
+
+There will be an optional parameter to `st.connection` to provide the full (user + base) token to make the RCR connection if users wish to implement something custom.
```
This is planned to be a build-in connection object in Streamlit, right?

sfc-gh-jkinkead replied:
Correct.

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,170 @@
+### Other notes
+All `ttl` and `max_entries` cache evictions happen only on write, not on read. This is how the underlying `TTLCache` works. This will be fine for session-scoped items, since they’ll be evicted manually when the session expires, but needs to be called out in the docs as a limitation of the existing cache. This will really only matter for `ttl` when users treat it as a guaranteed shutdown, and not  ...
+
+This implementation also will have `max_entries` scoped to the session, not scoped globally. This can be documented, and shouldn’t be an issue for connections, since they will not grow beyond the number of active sessions.
```
Is this possible to do with `TTLCache` or does it need a custom implementation or dedicated `TTLCache` per session?

sfc-gh-jkinkead replied:
See the link under "Eviction implementation". `TTLCache` is designed to be extended in order to implement eviction hooks. These would still be global cached - but with a session ID as a part of the cache key.

sfc-gh-jkinkead committed:
Cleanup based on feedback.

* Rename parameters for clarity.
* Use literal type instead of boolean for `scope` to allow new values in the figure.
* Expand SPCS in a few places, since this term isn't used much in public docs.
* Fix typo.

sfc-gh-jkinkead reviewed (commented):
Comments for @jrieke -

> session_scoped: bool = False on st.cache_resource: I think we should do this as scope="global"|"session" ...

Agreed, updated.

> eviction_hook on st.cache_resource ...

I want it to be crystal clear to users that this hook is called when something is removed from the cache, not for any other reason. Words like "cleanup", "teardown", and "close" all imply that this could be called for reasons other than cache eviction.

I do like the `on_` prefix though. I used Lukas' suggestion of `on_expire`.

> Not an extremely strong opinion but I think on st.connection we should use the same parameter names as on st.cache_resource.

Agreed that we should keep things consistent where it makes sense; see below.

> the shutdown function you propose above is not different from the eviction_hook function, correct?

It is semantically. Shutdown is an action that performs a shutdown, which you might do for a number of reasons; an eviction hook is a function that's called on cache eviction. A shutdown is one thing you might use an eviction hook for - but you could also use an eviction hook for any number of other things. I think reusing the name here is confusing.

> "This will NOT be called when an app is shut down for global resources." -> This means when I kill the Streamlit server the eviction_hook function will not be called? Does this create any problems, e.g. for a database connection, because something on the side of the database is kept open? Or is killing the Python process anyway equivalent to shutting down the connection?

There's a few things here:

1. It's generally hard to guarantee that code runs when an application terminates. It's fairly easy to set up triggers - but it's really easy to have them get bypassed in situations like container death or out-of-memory situations. I would like us to avoid opening that can of worms (bug reports will be filed if we write this).
2. Live connections using a socket (traditional database connections) will usually clean up pretty quickly when a connection drops on the client side. Snowflake is an exception, since it's RESTful, but does have session timeouts which clean things up eventually.
3. Libraries will generally try to clean things up on shutdown where feasible.

I think if we _do_ want global shutdown hooks, it would be better to implement them as a part of the core Streamlit library rather than the caching layer, since that's the right place to execute them. You really want them to run just before the process exits, after the HTTP server and all user script execution has stopped.

> The proposed snowflake-rcr connection:

> Is that something that's usable outside of a SiS app?

Yes, it would work for _any_ Streamlit running on SPCS with [caller's rights](https://docs.snowflake.com/en/developer-guide/snowpark-container-services/additional-considerations-services-jobs#label-spcs-additional-considerations-configuring-callers-rights) enabled on the service.

> Would the dev explicitly call st.connection("snowflake-rcr") in their SiS app, or are we calling this automatically if the app is using caller's rights?

Yes. This is a per-connection setting; that's how it works on SPCS.

> I think what I'm trying to get to: is this something we want to have public and distributed as part of the Streamlit repo, or is this something that should be private and potentially not even ship with Streamlit, since it's a SiS-only feature?

Per above it's an SPCS feature. Users would be able to enable it in _any_ SPCS environment.

> If it should be public, we should also think about renaming it (e.g. snowflake-callers-rights) since we usually try to avoid acronyms in Streamlit. Or should this even be a parameter on the existing st.connection("snowflake")?

Making it be a parameter is interesting, actually. I like that idea. The rest of the design still works - but it makes the user API more direct. I've incorporated that.

sfc-gh-jkinkead committed:
More fixes.

Fix some typos and a few wording inconsistencies.

Be more clear about TTLCache.

lukasmasuch commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,178 @@
+## Summary
+
+This proposal adds support for session-scoped values in `st.cache_resource` and `st.connection`. It also proposes using
+this new support to implement a new connection type, `snowflake-rcr`, which handles creating [restricted caller's rights sessions](https://docs.snowflake.com/en/developer-guide/snowpark-container-services/additional-considerations-services-jobs#label-spcs-additional-considerations-configuring-callers-rights) when running in Snowpark Container Services.
```
nit: I think `snowflake-rcr` might be a bit misleading here since as far as I understand, we will just use `st.connect("snowflake")` with additional flags that activate this behaviour.

sfc-gh-jkinkead replied:
Good catch! I changed this elsewhere in the document but didn't notice this part. I'll update.

The new proposal is to have a new parameter to toggle on caller's rights.

lukasmasuch reviewed (approved):
LGTM 👍 One comment about `shutdown`. But specs will need to approvals, so probably waiting for @jrieke to get this merged in

sfc-gh-jkinkead committed:
Fix summary.

sfc-gh-jkinkead committed:
s/shutdown/close/g

jrieke commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,181 @@
+
+### API Changes
+
+`st.cache_resource` and `st.cache_data` will have two new parameters:
```
For `st.cache_data`, I would only add `scope` but not `on_expire`. Can't really see what the use cases for `on_expire` with cached data are, or do you have any use cases in mind?

sfc-gh-jkinkead replied:
> do you have any use cases in mind?

I feel like we were pushing for full symmetry when we discussed this as a group. I can't really think of a good use for this.

When starting to implement this, I realized quickly that the data cache implementation is quite a bit more complicated than the resource cache. The interaction between eviction and varying cache persistence layers would need to be fleshed out more anyway - so I was planning on removing this feature anyway.

Pulled from the spec!

jrieke commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,181 @@
+        expired item will be provided to the function as an argument.
+
+        This is only useful for caches which will expire entries normally: Those with
+        ``max_entries`` or ``ttl`` settings, or those using ``scope="session"``.
```
I guess if you call `st.cache_resource.clear()` or `cached_function.clear()`, this would also fire? (At least I think it should!). 

1. Would be great to document that here, or maybe even turn this into an explicit bullet point list of when the function is called.

2. If that's the case, then I think the name `on_expire` is a bit confusing. Understand the desire not to call it anything like `close` or `teardown` though since we're not calling this when shutting down the server. Some more ideas:

- `on_release` -> might work well for resources
- `on_evict` -> still don't love it but at least more correct than `on_expire`

sfc-gh-jkinkead replied:
> I guess if you call st.cache_resource.clear() or cached_function.clear(), this would also fire? (At least I think it should!).

Yes! These are considered evictions.

> Would be great to document that here

Done.

> or maybe even turn this into an explicit bullet point list of when the function is called

I think this might be better done in the docs site itself in a cache how-to section.

> I think the name on_expire is a bit confusing

I renamed to `on_evict`.

sfc-gh-jrieke replied:
Had another discussion with ChatGPT and it recommended `on_release` over `on_evict` because `on_evict` sounds a bit weird with manual cache clearing + too jargony. I think I agree with that, would go with `on_release`, is that ok for you?

sfc-gh-jkinkead replied:
WFM! Updated.

jrieke commented on specs/0001-session-scoped-connections-and-rcr/product-spec.md:
```diff
@@ -0,0 +1,181 @@
+
+Note that the token reading here is different from the `snowflake` connection, which relies on the driver to do all internals. We can’t do that here, as the driver doesn’t natively support these RCR tokens.
+
+`st.connection` will have two new optional parameters read by the `"snowflake"` connection type.
```
Are these supposed to be a) normal keyword arguments on `st.connection` or b) parameters you usually define in `secrets.toml` for the connection, and can overwrite them via `**kwargs` in `st.connection`?

If (a), I'm not a fan of that, because then we're adding two very Snowflake-specific parameters to `st.connection`, which by itself isn't Snowflake-specific. In that case I would rather go back to adding a new connection type, e.g. `st.connection("snowflake-callers-rights")`. If it's (b), I'd be fine with that, even though I could see that it's not super discoverable (so maybe a separate connection type is better either way?).

sfc-gh-jkinkead replied:
> a) normal keyword arguments

Not sure what you mean by "normal". `connection_factory` (what backs `st.connection`) [has only a handful of named parameters](https://github.com/streamlit/streamlit/blob/7af8313983a03120f282daaba4e4c6324c2e23bd/lib/streamlit/runtime/connection_factory.py#L225-L273), but passes all other arguments (`**kwargs`) down to the base connection. The intent is that these new parameters will be part of this `kwargs`.

> b) parameters you usually define in secrets.toml for the connection

I hadn't considered these something that users would override in `secrets.toml`. Whether or not you're using caller's rights should be specific to your code path, not your config - and furthermore if you _did_ have a `secrets.toml` for the connection, you'd likely want everything to be the same except for the connection mode for the caller's rights mode.

Anyway, I think the short answer is b) is what I intended. I think either a connection type or parameter is fine by me. As far as discoverability goes, I think this just needs to be documented well. People aren't going to be navigating the source code for registered connection types in order to figure out how to connect to Snowflake, so class vs. parameter shouldn't matter too much. The advantage to having a shared class is the docs can all be in one place; the disadvantage is there won't be a separate docs page automatically for caller's rights.

Let me know what you think! Either works with me. I originally had a separate name because a named connection feels _really_ obvious from a calling perspective, but a parameter works better if you want to have a shared `secrets.toml` entry for both (which you generally would).

sfc-gh-jrieke replied:
"Normal" = appears as a parameter in the docstring of `st.connection` (i.e. [here](https://docs.streamlit.io/develop/api-reference/connections/st.connection) – seems like this isn't what you want anyway, right?

OK yeah I think I'm also fine either way, especially because I don't understand all the details of the `SnowflakeConnection` class :D If we're going with parameters (which does seem slightly cleaner), I think it might be a good idea to be able to also set them through `secrets.toml` (we do the same thing for auth, you can set pretty much everything through `secrets.toml` and (in the future) also through `st.login`). I know it's a bit muddy between secrets and config then, but often it's just easier to have everything in one place. But defer to you on the specifics here!

sfc-gh-jkinkead replied:
Yeah, I'll have a better opinion on secrets once I implement that part - it'll just be really awkward for the caller's rights setting to have it overridden. It doesn't mean we shouldn't add it for consistency - I just think it won't be a good way to do it.

jrieke reviewed (commented):
Thanks for all the replies, very helpful! Agree with the general direction of not calling this when we shut down the server. Added a few more small comments on the API though. The other big thing is the new connection type, I think that needs a bit of additional clarification.

sfc-gh-jkinkead committed:
Rename on_expire back to on_evict.

Remove eviction hooks from cache_data.

sfc-gh-jrieke reviewed (approved):
Approving, but please change `on_evict` to `on_release`, or let me know if you have a strong opinion here :)

sfc-gh-jkinkead committed:
Update OnEvict to OnRelease.

sfc-gh-jkinkead committed:
Merge branch 'develop' into jkinkead-session-scoped-connections
